### PR TITLE
Elektra Web - docker images

### PR DESCRIPTION
**Pull request by omnidan**
I created (and pushed) the following docker images (as per #1901):

- `elektra/web-base` - base image for elektra web, all other images build upon this (builds elektra with `yajl` and `kdb`)
- `elektra/elektrad` - image that only starts elektrad
- `elektra/webd` - image that only starts webd
- `elektra/elektrad-demo` - same as `elektrad`, but with a KDB config set up (for demo, should run on http://elektrad-demo.libelektra.org)
- `elektra/webd-demo` - same as `webd`, but with 2 instances already created, they both connect to http://elektrad-demo.libelektra.org with different visibility levels (for demo, should run on http://webui.libelektra.org)
- `elektra/web` - image that starts elektrad & webd (for those who just want to try out Elektra Web locally, also mentioned in quickstart in README of #2099)

All images are available under the `1.5` and `latest` tags.

You can find the images on Docker Hub: https://hub.docker.com/u/elektra/

@markus2330 @ingwinlu please let me know your Docker Hub usernames so that I can add you as owners to the `elektra` group

### CODE REVIEW

File: scripts/docker/web/web/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FROM elektra/web-base:latest
+
+WORKDIR /home/elektra
+USER elektra
+
+# create start script
+RUN printf "#!/bin/bash\nkdb run-elektrad &\nkdb run-web" > start
```
Comment (markus2330) [inline scripts/docker/web/web/Dockerfile]: I thought you will split this up into two containers?

Please give a description about the individual dockerfiles. (as comments and in the README.md)

Comment (omnidan) [reply scripts/docker/web/web/Dockerfile]: @markus2330 I explained this in the PR description, `elektra/web` contains both for testing purposes (quickstart guide in README)

Comment (omnidan) [reply scripts/docker/web/web/Dockerfile]: I added the description to a README file

Comment (markus2330) [reply scripts/docker/web/web/Dockerfile]: Thank you! Please always keep all information within the repo (and not in the PR description).

**Comment by markus2330** [comment]
And please do not forget to update release notes.

**Comment by omnidan** [comment]
@markus2330 the docker images are already mentioned in the #2099 PR

**Comment by markus2330** [comment]
They need to be in every PR individually so that the build server passes.

**Comment by omnidan** [comment]
@markus2330 alright, I added them here too

**Comment by markus2330** [comment]
@ingwinlu Can you review this please?

Would be great if the demo is updated to the latest version.

File: scripts/docker/web/base/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+# elektra web deps
+RUN apt-get install -y libyajl-dev curl nodejs-legacy npm
+RUN curl -sL https://deb.nodesource.com/setup_10.x | bash -
+RUN apt-get install -y nodejs
+
+# pull latest libelektra
+RUN mkdir /home/elektra
+WORKDIR /home/elektra
+RUN git clone https://github.com/omnidan/libelektra.git
```
Comment (ingwinlu) [inline scripts/docker/web/base/Dockerfile]: Don't you want to use ElektraInitiative's repo here?
you also want git clone --branch my_abc http://git.abc.net/git/abc.git and probably a shallow clone.

File: scripts/docker/web/elektrad-demo/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+FROM elektra/web-base:latest
+
+WORKDIR /home/elektra
+
+# prepare demo environment
+
+# mount copy of /etc/hosts to user/hosts
+
+# mount as root user
+RUN kdb mount --with-recommends hosts user/hosts hosts
```
Comment (ingwinlu) [inline scripts/docker/web/elektrad-demo/Dockerfile]: will that work? shouldn't you be the elektra user when inheriting from your base image?

File: scripts/docker/web/elektrad-demo/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+# create user/app structure
+RUN kdb set user/app
+
+RUN kdb set user/app/peers
+RUN kdb set user/app/peers/#0 "192.168.0.43"
+RUN kdb setmeta user/app/peers/#0 check/validation "^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$"
+RUN kdb setmeta user/app/peers/#0 check/validation/message "invalid ip address"
+RUN kdb setmeta user/app/peers/#1 check/validation "^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$"
+RUN kdb setmeta user/app/peers/#1 check/validation/message "invalid ip address"
+RUN kdb set user/app/peers/#1 "192.168.0.44"
```
Comment (ingwinlu) [inline scripts/docker/web/elektrad-demo/Dockerfile]: what are the ip addresses used here?

Comment (omnidan) [reply scripts/docker/web/elektrad-demo/Dockerfile]: they are just an example for the demo, they have no meaning

File: scripts/docker/web/elektrad-demo/Dockerfile (hunk 3 above)
Comment (omnidan) [reply scripts/docker/web/elektrad-demo/Dockerfile]: yes, `USER` is not called in the base image

File: scripts/docker/web/base/Dockerfile (hunk 2 above)
Comment (omnidan) [reply scripts/docker/web/base/Dockerfile]: you are right, I will do that now that the other PR is merged and rebuild the images

File: scripts/docker/web/elektrad-demo/Dockerfile (hunk 4 above)
Comment (ingwinlu) [reply scripts/docker/web/elektrad-demo/Dockerfile]: in that case: you could provide a file containing the demo data and mount that inside the container. would reduce image layers by a lot and is probably easier to maintain in the long run (maybe you can even use something that is around in the repository for tests)

File: scripts/docker/web/base/Dockerfile (hunk 2 above)
Comment (ingwinlu) [reply scripts/docker/web/base/Dockerfile]: I like to get .tar.gz archives from github. This removes the need to install git and it makes it very easy to insert a build-arg that represents the version I want to build.

**Comment by omnidan** [comment]
I updated, rebuilt & pushed all of the images. Thanks for the review and recommendations @ingwinlu

**Comment by markus2330** [comment]
`docker run -d -it -p 33333:33333 -p 33334:33334 elektra/web` does not download anything new. Am I doing something wrong?

File: scripts/docker/web/base/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+# elektra web deps
+RUN apt-get install -y libyajl-dev curl nodejs-legacy npm
+RUN curl -sL https://deb.nodesource.com/setup_10.x | bash -
+RUN apt-get install -y nodejs
+
+# pull latest libelektra
+RUN mkdir /home/elektra
+WORKDIR /home/elektra
+RUN git clone --depth 1 https://github.com/ElektraInitiative/libelektra.git
```
Comment (ingwinlu) [inline scripts/docker/web/base/Dockerfile]: You should pin a version for reproducible builds.

Comment (ingwinlu) [reply scripts/docker/web/base/Dockerfile]: you did this before by checking out the tag. I would go for a build-arg that you feed to a --branch parameter of clone (you can let it default to master if you want).

**Comment by omnidan** [comment]
@markus2330 can you try `docker run -d -it -p 33333:33333 -p 33334:33334 elektra/web:latest` or `docker run -d -it -p 33333:33333 -p 33334:33334 elektra/web:1.5` and see if that makes a difference?

**Comment by ingwinlu** [comment]
@markus2330 add the --pull flag

File: scripts/docker/web/base/Dockerfile (hunk 5 above)
Comment (omnidan) [reply scripts/docker/web/base/Dockerfile]: @ingwinlu good idea, I did that now!

**Comment by omnidan** [comment]
@ingwinlu I pinned the version and added a build argument for it

**Comment by markus2330** [comment]
> see if that makes a difference?

elektra/web:1.5 downloads something but how can I know the difference? It always shows "v1.5" and rename works.

**Comment by markus2330** [comment]
Thank you for the review and the new dockerfiles!

**Comment by omnidan** [comment]
@markus2330 if it shows `v1.5` in the UI you should already have the latest version

**Pull request merged**